src/shared/wizards/util.ts

```diff
@@ -0,0 +1,158 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// This file is mostly for keeping telemetry code separate from the core wizard code
+import * as _ from 'lodash'
+import { ControlSignal } from './stateController'
+
+interface TraceData {
+    prop: string
+    result: any
+    duration: number
+    totalSteps: number
+    currentStep: number
+}
+
+interface ResultBase {
+    type: 'Completed' | 'Failed' | 'Cancelled'
+    duration: number
+    uniqueSteps: number
+    totalPrompts: number
+}
+
+interface TraceCompleted<TState> extends ResultBase {
+    type: 'Completed'
+    state: TState
+}
+
+interface TraceFailed extends ResultBase {
+    type: 'Failed'
+    reason: string | Error
+    prop: string
+}
+
+interface TraceCancelled extends ResultBase {
+    type: 'Cancelled'
+    prop: string
+    furthestStep: string
+}
+
+interface TraceReporter {
+    stop: (result: any) => void
+}
+
+export type WizardTraceResult<TState> = TraceCompleted<TState> | TraceFailed | TraceCancelled
+
+export class WizardTrace<TState> {
```

Comment:
This will be used to inject metadata into telemetry. It can capture a lot of information:
* Total duration
* Time spent per-prompt
* What step did the user cancel on?
* Was there an uncaught error? If so, which step?
* What was the furthest step the user got to?
* How many total prompts did the user see?

I doubt everything will be used, but it's a start.

Reply:
Yeah I was going to pull it out, though I wanted to highlight it to show some ideas I had in mind. The changes to the wizard are very minor, but I did need to implement this on my 'main' branch to get an idea of how it'd work.

The step number _does_ mean something, though you're correct that a step identifier is needed. In this case the default identifier (currently implemented) is just the bound property which is easily human readable but is more tightly coupled to the code. Currently just thinking of having a required 'id' field somewhere.